File: src/C/new_main.cpp
#include "orbit_integrator.hpp"
#include "jeans_initial.hpp"
#include <iostream>
#include <fstream>
#include <string>
#include <tuple>
#include <cmath>
#include <iomanip>

int flatten_index(const IndexTriple& idx, int N) {
    return idx.x * N * N + idx.y * N + idx.z;
}

int main() {
    // === Simulation Parameters ===
    const int    N           = 128;
    const double box_size    = 1.0;
    const int    N_particles = 5000;
    const double dt          = 2e-4;
    const int    n_steps     = 500;
    const std::string dp     = "ngp";
    const std::string solver = "periodic";
    const std::string integrator = "dkd";
    const std::string mode   = "expand";
    const double a           = 0.05;
    const double G           = 1.0;

    // === Initialization ===
    ParticleArray positions, velocities;
    MassArray    masses;
    std::tie(positions, velocities, masses)
        = create_particles_single(N_particles, box_size, a, 1.0, mode, solver, G);//,true,0.4);
    std::cout<<"HO\n";
    // Open CSV outputs
    std::ofstream log_file("Result/simulation_output_"+dp+""+solver+""+integrator+"_"+mode+".csv");
    log_file << std::scientific << std::setprecision(17);
    log_file << "step,KE,PE,Total,Px,Py,Pz\n";

    std::ofstream pos_file("Result/particle_positions_"+dp+""+solver+""+integrator+"_"+mode+".csv");
    pos_file << "step,id,x,y,z\n";

    // Clear any old potential dump
    {
        std::ofstream clear_file("potential_N128.csv", std::ios::trunc);
    }

    // Precompute Green's function for isolated solver
    std::vector<std::complex<double>> G_k;
    if (solver == "isolated") {
        G_k = compute_green_ft(N, box_size);
    }
    double KE = 0.0;
    std::array<double,3> P = {0.0, 0.0, 0.0};
    std::cout<<"Before Update"<<"\n";
    for (size_t i = 0; i < positions.size(); ++i) {
        const auto& v = velocities[i];
        KE += 0.5 * masses[i] * (v[0]*v[0] + v[1]*v[1] + v[2]*v[2]);
        for (int d = 0; d < 3; ++d) {
            P[d] += masses[i] * v[d];
        }
    }
    auto [phi_flat, weights_list]
    = compute_phi(positions, masses, N, box_size, dp, "periodic", G_k);

    double PE = 0.0;
    for (size_t i = 0; i < positions.size(); ++i) {
        double phi_p = 0.0;
        for (auto& [idx, w] : weights_list[i]) {
            int flat = flatten_index(idx, N);
            phi_p += w * phi_flat[flat];
        }
        PE += masses[i] * phi_p;
    }
    PE *= 0.5;  // account for double‐counting

    double total = KE + PE;

    // 5) Log to console & CSV
    std::cout   << "  KE="    << KE
                << ", PE="   << PE
                << ", Total="<< total
                << ", Px="   << P[0]
                << ", Py="   << P[1]
                << ", Pz="   << P[2]
                << "\n";




    // === Time‐integration loop ===
    for (int step = 0; step < n_steps; ++step) {
        // 1) Advance one step
        StepResult step_result;
        if      (integrator == "kdk") step_result = kdk_step (positions, velocities, masses, dt, N, box_size, dp, solver, G_k);
        else if (integrator == "dkd") step_result = dkd_step (positions, velocities, masses, dt, N, box_size, dp, solver, G_k);
        else /*rk4*/                  step_result = rk4_step(positions, velocities, masses, dt, N, box_size, dp, solver, G_k, G);

        // 2) Unpack the survivors
        positions = std::move(step_result.positions);
        velocities = std::move(step_result.velocities);
        masses    = std::move(step_result.masses);
        size_t M = positions.size();  // actual number of particles remaining

        // 3) Kinetic energy and total momentum
        double KE = 0.0;
        std::array<double,3> P = {0.0, 0.0, 0.0};
        for (size_t i = 0; i < M; ++i) {
            const auto& v = velocities[i];
            KE += 0.5 * masses[i] * (v[0]*v[0] + v[1]*v[1] + v[2]*v[2]);
            for (int d = 0; d < 3; ++d) {
                P[d] += masses[i] * v[d];
            }
        }

        // 4) Potential energy via PM
        auto [phi_flat, weights_list]
            = compute_phi(positions, masses, N, box_size, dp, solver, G_k);

        double PE = 0.0;
        for (size_t i = 0; i < M; ++i) {
            double phi_p = 0.0;
            for (auto& [idx, w] : weights_list[i]) {
                int flat = flatten_index(idx, N);
                phi_p += w * phi_flat[flat];
            }
            PE += masses[i] * phi_p;
        }
        PE *= 0.5;  // account for double‐counting

        double total = KE + PE;

        // 5) Log to console & CSV
        std::cout << "Step " << step
                  << "  KE="    << KE
                  << ", PE="   << PE
                  << ", Total="<< total
                  << ", Px="   << P[0]
                  << ", Py="   << P[1]
                  << ", Pz="   << P[2]
                  << "\n";

        log_file << step << "," << KE << "," << PE << "," << total
                 << "," << P[0] << "," << P[1] << "," << P[2] << "\n";

        // 6) Dump surviving particle positions
        for (size_t i = 0; i < M; ++i) {
            pos_file << step << "," << i << ","
                     << positions[i][0] << ","
                     << positions[i][1] << ","
                     << positions[i][2] << "\n";
        }

        // 7) Record the mid‐plane potential slice
        std::ofstream pot_file("potential_N128.csv", std::ios::app);
        int z = N/2;
        for (int y = 0; y < N; ++y) {
            for (int x = 0; x < N; ++x) {
                int flat_idx = x*N*N + y*N + z;
                pot_file << phi_flat[flat_idx] << " ";
            }
            pot_file << "\n";
        }
    }

    log_file.close();
    pos_file.close();
    return 0;
}

File: src/python/new_main.py
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.animation as animation
from matplotlib.animation import PillowWriter
from new_orbit_integrator import kdk_step, dkd_step,rk4_step
from new_orbit_integrator import compute_phi
from mpl_toolkits.mplot3d import Axes3D
from jeans_initial import create_particles_single, create_particles_double
from poisson_solver import green

# === Simulation parameters ===
N = 128                    # Grid size: N x N x N
box_size = 1.0
N_particles =  100         # 10000
center = N // 2
dt = 2e-4
n_steps = 100              # 200
dp = 'cic'                 # 'ngp', 'cic', or 'tsc'
solver = 'periodic'        # 'isolated', 'periodic 
integrator = 'dkd'         # 'kdk' or 'dkd' or 'rk4' 
dx = box_size/N
mode = 'stable'
a = 0.01

if solver == 'isolated':
    G_k = green(N, box_size)
else:
    G_k =0

# === useful function ===
def potential_slice(phi, box_size, title):
    """Plot mid-plane slice of potential."""
    plt.imshow(phi[:,:,center], extent=[0, box_size, 0, box_size], origin='lower')
    plt.colorbar(label="Potential Φ")
    plt.title(title)
    plt.xlabel("x")
    plt.ylabel("y")
    plt.show()

def particle_slice_XY(positions, title):
    fig_init, ax_init = plt.subplots()
    ax_init.scatter(positions[:, 0], positions[:, 1], s=5, alpha=0.6)
    ax_init.set_aspect('equal')
    ax_init.set_xlim(0, box_size)
    ax_init.set_ylim(0, box_size)
    ax_init.set_xlabel("x")
    ax_init.set_ylabel("y")
    ax_init.set_title(f'Particle SLice_{title}(XYplane)')
    plt.tight_layout()
    plt.show()

def compute_total_momentum(velocities, masses):
    """Compute momentum (vector)"""
    return np.sum(masses[:, None] * velocities, axis=0)

def compute_total_energy(positions, velocities, masses, N, box_size,dp,solver, G_k):
    """Compute total energy (kinetic + potential) of the system."""
    KE = 0.5 * np.sum(masses * np.sum(velocities**2, axis=1))

    phi,weights = compute_phi(positions, masses, N, box_size, dp, solver, G_k)
    particle_values = []
    for weight in weights:
        phi_par = 0.0  # use scalar
        for idx, w in weight:
            phi_par += phi[idx] * w
        particle_values.append(phi_par)
    particle_values = np.array(particle_values)
    PE = 0.5*np.sum(masses*particle_values)
    return KE, PE, KE+PE

def compute_energies_direct(x, v, m, G=1.0, softening=1e-5):
    #can be removed
    N = len(m)
    KE = 0.5 * np.sum(m * np.sum(v**2, axis=1))
    PE = 0.0
    for i in range(N):
        for j in range(i+1, N):
            dx = x[i] - x[j]
            r2 = np.dot(dx, dx) + softening**2
            PE -= G * m[i] * m[j] / np.sqrt(r2)
    return KE, PE, KE + PE


# === main simulation ===
def main():
    np.random.seed(42)
    # single plummer(mode = 'stable, contract, expand)
    positions, velocities, masses = create_particles_single(N_particles,box_size,a = a, M = 1.0, mode = mode ,solver = solver,G=1.0)
    # two plummer(mode, addv= True,False, v_offset=)
    #positions, velocities, masses = create_particles_double(N_particles,box_size,a = a, M = 1.0, mode=mode,solver = solver,G=1.0 , add_initial_velocity=True, v_offset= 0.1)
    particle_slice_XY(positions, 'Initial setup')
    # initial total energy/momentum
    # K_direct, P_direct, energy_direct = compute_energies_direct(positions, velocities, masses)
    K_poi, P_poi, energy_poisson = compute_total_energy(positions,velocities,masses, N, box_size, dp , solver,G_k)
    momentum = compute_total_momentum(velocities,masses)
    # Q_J_dir = 2 * K_direct / abs(P_direct)
    Q_J_poi = 2*K_poi/(abs(P_poi))
    # print(f"Initial Energy(direct) = {energy_direct:.4f}, Qj = {Q_J_dir:.4f}")
    print(f"Initial Energy(poisson) = {energy_poisson:.4f},Qj = {Q_J_poi:.4f}")
    print("Initial momentum =", momentum)

    # initial setup figure
    particle_slice_XY(positions, 'Initial setup')

    # lists for update date
    momentum_errors = []         # total error ||ΔP||
    momentum_errors_xyz = []     # per-axis error [|ΔPx|, |ΔPy|, |ΔPz|]

    frames = []           # potential field frames
    particle_frames = []  # particle position frames
    energies = []
    energies.append([K_poi,P_poi])
    saved_frames = set()

    # update
    for step in range(n_steps):
        if integrator == 'kdk':
            positions, velocities, masses, phi = kdk_step(positions, velocities, masses, dt, N, box_size, dp, solver,G_k)
        elif integrator == 'dkd':
            positions, velocities,masses, phi = dkd_step(positions, velocities, masses, dt, N, box_size, dp, solver,G_k)
        elif integrator == 'rk4':
            positions, velocities,masses, phi = rk4_step(positions, velocities, masses, dt, N, box_size, dp, solver,G_k)

        # compute energy/momentum each step
        KE, PE, total_energy = compute_total_energy(positions, velocities, masses, N, box_size,dp,solver,G_k)
        energies.append([KE, PE])
        current_momentum = compute_total_momentum(velocities, masses)
        delta_P = current_momentum - momentum
        momentum_errors.append(np.linalg.norm(delta_P))
        momentum_errors_xyz.append(np.abs(delta_P))

        # Save frames every 5 steps
        if step % 1 == 0:
            frames.append(phi[:,:,center].T.copy())
            particle_frames.append(positions.copy())
    
    PP_anim(frames, particle_frames, box_size, dp, solver)
    energy(energies)
    momentum_plot(momentum_errors_xyz)
        
# === figures ===
def PP_anim(frames, particle_frames, box_size,dp,solver):
    fig3, ax3 = plt.subplots()
    im = ax3.imshow(frames[0], extent=[0, box_size, 0, box_size], origin='lower',
                    vmin=np.min(frames[0]), vmax=np.max(frames[0]), cmap='viridis')
    cbar_potential = plt.colorbar(im, ax=ax3)    # add color bar
    cbar_potential.set_label("Gravitational Potential")
    scatter = ax3.scatter(particle_frames[0][:,0], particle_frames[0][:,1], s=1, color='white')

    ax3.set_xlim(0, box_size)
    ax3.set_ylim(0, box_size)
    ax3.set_xlabel("x")
    ax3.set_ylabel("y")
    title3 = ax3.set_title(f"Particles + Potential at Frame 0")
    plt.tight_layout()

    def animate_combined(i):
        im.set_data(frames[i])
        scatter.set_offsets(particle_frames[i])
        title3.set_text(f"Particles + Potential at Frame {i}")
        return im, scatter, title3
    
    ani_combined = animation.FuncAnimation(fig3, animate_combined, frames=len(frames), interval=200, blit=False)
    # save gif
    ani_combined.save(f"PP_anim_{dp}_{solver}.gif", writer='pillow')
    plt.show()

def energy(energies):
    energies = np.array(energies)
    KEs = energies[:,0]
    PEs = energies[:,1]
    TEs = KEs + PEs

    plt.figure()
    plt.plot(KEs, label="Kinetic Energy")
    plt.plot(PEs, label="Potential Energy")
    plt.plot(TEs, label="Total Energy")
    plt.xlabel("Step")
    plt.ylabel("Energy")
    plt.title("Energy Conservation Test")
    plt.legend()
    plt.grid()
    plt.show()

    plt.figure()
    plt.plot(TEs-TEs[0], label="Delta Total Energy")
    plt.xlabel("Step")
    plt.ylabel("Energy diff")
    plt.title("Energy Conservation Test")
    plt.legend()
    plt.grid()
    plt.show()

def momentum_plot(momentum_errors_xyz):
    momentum_errors_xyz = np.array(momentum_errors_xyz)  # shape: (n_steps, 3)

    plt.figure()
    plt.plot(momentum_errors_xyz[:, 0], label='|ΔP_x|')
    plt.plot(momentum_errors_xyz[:, 1], label='|ΔP_y|')
    plt.plot(momentum_errors_xyz[:, 2], label='|ΔP_z|')
    plt.xlabel("Step")
    plt.ylabel("Momentum Error")
    plt.title("Momentum Conservation Error per Component")
    plt.legend()
    plt.grid()
    plt.show()

if __name__ == "__main__":
    main()
